cloudinit/tests/test_features.py

```diff
@@ -0,0 +1,59 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+# pylint: disable=no-member,no-name-in-module
+"""
+This file is for testing the feature flag functionality itself,
+NOT for testing any individual feature flag
+"""
+import pytest
+from importlib import reload
+from pathlib import Path
+
+import cloudinit
+import cloudinit.features
+
+
+@pytest.yield_fixture()
+def create_override(request):
+    override_path = Path(cloudinit.__file__).parent / 'feature_overrides.py'
+    with override_path.open('w') as f:
+        for key, value in request.param.items():
+            f.write('{} = {}\n'.format(key, value))
+
+    # The goal is to reload features.py so we also get the
+    # features_overrides.py stuff. This could normally be achieved with
+    # a single reload(cloudinit.features), however, since this fixture will be
+    # called multiple times in the same interpreter, calling
+    # reload(cloudinit.features) a second time won't refresh what already
+    # exists in feature_overrides.py because that's now already defined, so we
+    # need to call reload(cloudinit.feature_overrides) first. But the first
+    # time we call this fixture, cloudinit.feature_overrides doesn't exist yet,
+    # so we need to reload cloudinit.features first before we can reload
+    # cloudinit.feature_overrides, hence the three calls here.
+    reload(cloudinit.features)
+    reload(cloudinit.feature_overrides)
+    reload(cloudinit.features)
+
+    yield
+
+    override_path.unlink()
+
+
+class TestFeatures:
+    def test_feature_without_override(self):
+        from cloudinit.features import ERROR_ON_USER_DATA_FAILURE
+        assert ERROR_ON_USER_DATA_FAILURE is True
+
+    @pytest.mark.parametrize('create_override',
+                             [{'ERROR_ON_USER_DATA_FAILURE': False}],
+                             indirect=['create_override'])
+    def test_feature_with_override(self, create_override):
+        from cloudinit.features import ERROR_ON_USER_DATA_FAILURE
+        assert ERROR_ON_USER_DATA_FAILURE is False
```

Comment:
Indirectly parameterising a fixture that we are requesting in this test definition seems a little strange. I wonder if this is indicating that `create_override` might be better as a context manager? This would make this test definition something like:

```py
    def test_feature_with_override(self):
        with create_override({'ERROR_ON_USER_DATA_FAILURE': False}):
            from cloudinit.features import ERROR_ON_USER_DATA_FAILURE
            assert ERROR_ON_USER_DATA_FAILURE is False
```

(I think the only change required in `create_override` would be to wrap the `yield` in a `try`/`finally`, and to decorate it with [`contextlib.contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager).)

What do you think?

Reply:
I think of fixtures as a way of achieving a common setup/teardown for tests. I agree that the syntax is a little wonky, but the indirect parametrization is really just a way of passing arugments to a fixture. To me that seems more idiomatic (in a pytest sense) than rolling my own with context managers. That said, it's not a strongly held view, so if you'd really prefer I change it, let me know.